src/core/server/saved_objects/export/collect_exported_objects.ts

```diff
@@ -22,41 +23,75 @@ interface CollectExportedObjectOptions {
   /** The http request initiating the export. */
   request: KibanaRequest;
   /** export transform per type */
-  exportTransforms: Record<string, SavedObjectsExportTransform>;
+  typeRegistry: ISavedObjectTypeRegistry;
 }
 
 interface CollectExportedObjectResult {
   objects: SavedObject[];
+  excludedObjects: ExcludedObject[];
   missingRefs: CollectedReference[];
 }
 
+interface ExcludedObject {
+  id: string;
+  type: string;
+  reason?: string;
+}
+
 export const collectExportedObjects = async ({
   objects,
   includeReferences = true,
   namespace,
   request,
-  exportTransforms,
+  typeRegistry,
   savedObjectsClient,
 }: CollectExportedObjectOptions): Promise<CollectExportedObjectResult> => {
+  const exportTransforms = buildTransforms(typeRegistry);
+  const isExportable = buildIsExportable(typeRegistry);
+
   const collectedObjects: SavedObject[] = [];
   const collectedMissingRefs: CollectedReference[] = [];
+  const collectedNonExportableObjects: SavedObject[] = [];
   const alreadyProcessed: Set<string> = new Set();
 
   let currentObjects = objects;
   do {
-    const transformed = (
+    currentObjects = currentObjects.filter((object) => !alreadyProcessed.has(objKey(object)));
+
+    // first, evict current objects that are not exportable
+    const {
+      exportable: untransformedExportableInitialObjects,
+      nonExportable: nonExportableInitialObjects,
+    } = await splitByExportability(currentObjects, isExportable);
```

Comment:
See https://github.com/elastic/kibana/issues/99680#issuecomment-858370085 about the logic of the order.